File: rest-service/src/main/java/mb/lib/model/EmptyBlastQueryConfig.java
package mb.lib.model;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.veupathdb.lib.blast.BlastQueryConfig;
import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.field.ExpectValue;
import org.veupathdb.lib.blast.field.Location;
import org.veupathdb.lib.blast.field.QueryFile;

public class EmptyBlastQueryConfig extends EmptyBlastConfig implements BlastQueryConfig {

  public EmptyBlastQueryConfig(ObjectNode raw) {
    super(raw);
  }

  @Override
  public QueryFile getQueryFile() {
    return fromString(Flag.QueryFile, QueryFile::new);
  }

  @Override
  public void setQueryFile(String s) {
    raw.put(Flag.QueryFile, s);
  }

  @Override
  public Location getQueryLocation() {
    return get(Flag.QueryLocation, Location::fromJSON);
  }

  @Override
  public void setQueryLocation(Location location) {
    raw.set(Flag.QueryLocation, location.toJSON());
  }

  @Override
  public String getDBFile() {
    return getString(Flag.DBFile);
  }

  @Override
  public void setDBFile(String s) {
    raw.put(Flag.DBFile, s);
  }

  @Override
  public ExpectValue getExpectValue() {
    return fromString(Flag.ExpectValue, ExpectValue::new);
  }

  @Override
  public void setExpectValue(String s) {
    raw.put(Flag.ExpectValue, s);
  }

  @Override
  public Boolean getSoftMasking() {
    return getBoolean(Flag.SoftMasking);
  }

  @Override
  public void setSoftMasking(Boolean aBoolean) {
    raw.put(Flag.SoftMasking, aBoolean);
  }

  @Override
  public Boolean getLowercaseMasking() {
    return getBoolean(Flag.LowercaseMasking);
  }

  @Override
  public void setLowercaseMasking(Boolean aBoolean) {
    raw.put(Flag.LowercaseMasking, aBoolean);
  }

  @Override
  public String getEntrezQuery() {
    return getString(Flag.EntrezQuery);
  }

  @Override
  public void setEntrezQuery(String s) {
    raw.put(Flag.EntrezQuery, s);
  }

  @Override
  public Double getQueryCoverageHSPPercent() {
    return get(Flag.QueryCoverageHSPPercent, JsonNode::doubleValue);
  }

  @Override
  public void setQueryCoverageHSPPercent(Double aDouble) {
    raw.put(Flag.QueryCoverageHSPPercent, aDouble);
  }

  @Override
  public Long getMaxHSPs() {
    return getLong(Flag.MaxHSPs);
  }

  @Override
  public void setMaxHSPs(Long aLong) {
    raw.put(Flag.MaxHSPs, aLong);
  }

  @Override
  public Byte getDBSize() {
    return get(Flag.DBSize, s -> (byte) s.shortValue());
  }

  @Override
  public void setDBSize(Byte aByte) {
    raw.put(Flag.DBSize, aByte);
  }

  @Override
  public Short getSearchSpace() {
    return get(Flag.SearchSpace, JsonNode::shortValue);
  }

  @Override
  public void setSearchSpace(Short aShort) {
    raw.put(Flag.SearchSpace, aShort);
  }

  @Override
  public String getImportSearchStrategy() {
    return getString(Flag.ImportSearchStrategy);
  }

  @Override
  public void setImportSearchStrategy(String s) {
    raw.put(Flag.ImportSearchStrategy, s);
  }

  @Override
  public String getExportSearchStrategy() {
    return getString(Flag.ExportSearchStrategy);
  }

  @Override
  public void setExportSearchStrategy(String s) {
    raw.put(Flag.ExportSearchStrategy, s);
  }

  @Override
  public Double getExtensionDropoffUngapped() {
    return get(Flag.ExtensionDropoffUngapped, JsonNode::doubleValue);
  }

  @Override
  public void setExtensionDropoffUngapped(Double aDouble) {
    raw.put(Flag.ExtensionDropoffUngapped, aDouble);
  }

  @Override
  public Long getWindowSize() {
    return getLong(Flag.WindowSize);
  }

  @Override
  public void setWindowSize(Long aLong) {
    raw.put(Flag.WindowSize, aLong);
  }

  @Override
  public Boolean getRemote() {
    return getBoolean(Flag.Remote);
  }

  @Override
  public void setRemote(Boolean aBoolean) {
    raw.put(Flag.Remote, aBoolean);
  }
}

File: rest-service/src/main/kotlin/mb/lib/model/EmptyBlastQueryConfig.kt
package mb.lib.model

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.BlastQueryConfig
import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.field.ExpectValue
import org.veupathdb.lib.blast.field.Location
import org.veupathdb.lib.blast.field.QueryFile

class EmptyBlastQueryConfig(raw: ObjectNode): EmptyBlastConfig(raw), BlastQueryConfig {
  override var queryFile: QueryFile?
    get() = fromString(Flag.QueryFile, ::QueryFile)
    set(v) {
      raw.put(Flag.QueryFile, v?.value)
    }

  override var queryLocation: Location?
    get() = TODO("Not yet implemented")
    set(v) {
      raw.set<JsonNode>(Flag.QueryLocation, v?.toJSON())
    }

  override var dbFile
    get() = getString(Flag.DBFile)
    set(s) {
      raw.put(Flag.DBFile, s)
    }

  override var expectValue
    get() = fromString(Flag.ExpectValue, ::ExpectValue)
    set(s) {
      raw.put(Flag.ExpectValue, s?.value)
    }

  override var softMasking
    get() = getBoolean(Flag.SoftMasking)
    set(v) {
      raw.put(Flag.SoftMasking, v)
    }

  override var lowercaseMasking
    get() = getBoolean(Flag.LowercaseMasking)
    set(v) {
      raw.put(Flag.LowercaseMasking, v)
    }

  override var entrezQuery
    get() = getString(Flag.EntrezQuery)
    set(v) {
      raw.put(Flag.EntrezQuery, v)
    }

  override var queryCoverageHSPPercent
    get() = get(Flag.QueryCoverageHSPPercent, JsonNode::doubleValue)
    set(v) {
      raw.put(Flag.QueryCoverageHSPPercent, v)
    }

  override var maxHSPs
    get() = getLong(Flag.MaxHSPs)
    set(v) {
      raw.put(Flag.MaxHSPs, v)
    }

  override var dbSize
    get() = get(Flag.DBSize) { s -> s.shortValue().toByte() }
    set(v) {
      raw.put(Flag.DBSize, v?.toShort())
    }

  override var searchSpace: Short?
    get() = get(Flag.SearchSpace, JsonNode::shortValue)
    set(v) {
      raw.put(Flag.SearchSpace, v)
    }

  override var importSearchStrategy: String?
    get() = getString(Flag.ImportSearchStrategy)
    set(v) {
      raw.put(Flag.ImportSearchStrategy, v)
    }

  override var exportSearchStrategy: String?
    get() = getString(Flag.ExportSearchStrategy)
    set(v) {
      raw.put(Flag.ExportSearchStrategy, v)
    }

  override var extensionDropoffUngapped: Double?
    get() = get(Flag.ExtensionDropoffUngapped, JsonNode::doubleValue)
    set(v) {
      raw.put(Flag.ExtensionDropoffUngapped, v)
    }

  override var windowSize: Long?
    get() = getLong(Flag.WindowSize)
    set(v) {
      raw.put(Flag.WindowSize, v)
    }

  override var remote: Boolean?
    get() = getBoolean(Flag.Remote)
    set(v) {
      raw.put(Flag.Remote, v)
    }
}
